src/clj/puppetlabs/services/jruby/jruby_puppet_internal.clj

```diff
@@ -0,0 +1,205 @@
+  [timeout :- schema/Int
+   pool :- jruby-schemas/pool-queue-type]
+  (.pollFirst pool timeout TimeUnit/MILLISECONDS))
+
+(schema/defn borrow-from-pool!* :- (schema/maybe jruby-schemas/JRubyPuppetInstanceOrRetry)
+  "Given a borrow function and a pool, attempts to borrow a JRuby instance from a pool.
+  If successful, updates the state information and returns the JRuby instance.
+  Returns nil if the borrow function returns nil; throws an exception if
+  the borrow function's return value indicates an error condition."
+  [borrow-fn :- (schema/pred ifn?)
+   ;; it looks unusual that we accept both the pool and the pool-context
+   ;; as arguments, since the PoolContext contains a reference to a pool.  However,
+   ;; in cases such as a full pool flush operation, there may be two distinct
+   ;; pools in play, and the one in the pool-context may be different from the
+   ;; one we're borrowing from.
+   pool :- jruby-schemas/pool-queue-type
+   pool-context :- jruby-schemas/PoolContext]
+  (let [instance (borrow-fn pool)]
+    (cond (instance? PoisonPill instance)
+          (do
+            (.putFirst pool instance)
+            (throw (IllegalStateException.
+                     "Unable to borrow JRuby instance from pool"
+                     (:err instance))))
+
+          (jruby-schemas/jruby-puppet-instance? instance)
+          (do
+            (swap! (:state instance) (fn [m] (update-in m [:request-count] inc)))
+            instance)
+
+          ((some-fn nil? jruby-schemas/retry-poison-pill?) instance)
+          instance
+
+          :else
+          (throw (IllegalStateException.
+                   (str "Borrowed unrecognized object from pool!: " instance))))))
+
+(schema/defn ^:always-validate
+  borrow-from-pool :- jruby-schemas/JRubyPuppetInstanceOrRetry
+  "Borrows a JRubyPuppet interpreter from the pool. If there are no instances
+  left in the pool then this function will block until there is one available."
+  [pool-context :- jruby-schemas/PoolContext]
+  (borrow-from-pool!* borrow-without-timeout-fn
+                      (get-pool pool-context) pool-context))
+
+(schema/defn ^:always-validate
+  borrow-from-pool-with-timeout :- (schema/maybe jruby-schemas/JRubyPuppetInstanceOrRetry)
+  "Borrows a JRubyPuppet interpreter from the pool, like borrow-from-pool but a
+  blocking timeout is provided. If an instance is available then it will be
+  immediately returned to the caller, if not then this function will block
+  waiting for an instance to be free for the number of milliseconds given in
+  timeout. If the timeout runs out then nil will be returned, indicating that
+  there were no instances available."
+  [pool-context :- jruby-schemas/PoolContext
+   timeout :- schema/Int]
+  {:pre  [(>= timeout 0)]}
+  (borrow-from-pool!* (partial borrow-with-timeout-fn timeout)
+                      (get-pool pool-context) pool-context))
+
+(schema/defn ^:always-validate
+  return-to-pool
```

Comment:
Should this be `return-to-pool!` since it doesn't return anything but instead changes the argument? Not a big deal though, just like all my other comments :smile: